core/keyboardshortcut_registry.js

```diff
@@ -0,0 +1,351 @@
+ */
+Blockly.KeyboardShortcutRegistry.prototype.onKeyDown = function(workspace, e) {
+  var key = this.serializeKeyEvent_(e);
+  var shortcutNames = this.getKeyboardShortcuts(key);
+  if (!shortcutNames) {
+    return false;
+  }
+  for (var i = 0, shortcutName; (shortcutName = shortcutNames[i]); i++) {
+    var shortcut = this.registry_[shortcutName];
+    if (!shortcut.preconditionFn || shortcut.preconditionFn(workspace)) {
+      // If the key has been handled, stop processing shortcuts.
+      if (shortcut.callback && shortcut.callback(workspace, e, shortcut)) {
+        return true;
+      }
+    }
+  }
+  return false;
+};
+
+/**
+ * Gets the shortcuts registered to the given key code.
+ * @param {string} keyCode The serialized key code.
+ * @return {!Array<string>|undefined} The list of shortcuts to call when the
+ *     given keyCode is used. Undefined if no shortcuts exist.
+ * @public
+ */
+Blockly.KeyboardShortcutRegistry.prototype.getKeyboardShortcuts = function(
+    keyCode) {
+  return this.keyMap_[keyCode] || [];
+};
+
+/**
+ * Gets the serialized key codes that the shortcut with the given name is
+ * registered under.
+ * @param {string} shortcutName The name of the shortcut.
+ * @return {!Array<string>} An array with all the key codes the shortcut is
+ *     registered under.
+ * @public
+ */
+Blockly.KeyboardShortcutRegistry.prototype.getKeyCodeByShortcutName = function(
+    shortcutName) {
+  var keys = [];
+  for (var keyCode in this.keyMap_) {
+    var shortcuts = this.keyMap_[keyCode];
+    var shortcutIdx = shortcuts.indexOf(shortcutName);
+    if (shortcutIdx > -1) {
+      keys.push(keyCode);
+    }
+  }
+  return keys;
+};
+
+/**
+ * Serializes a key event.
+ * @param {!Event} e A key down event.
+ * @return {string} The serialized key code for the given event.
+ * @private
+ */
+Blockly.KeyboardShortcutRegistry.prototype.serializeKeyEvent_ = function(e) {
+  var serializedKey = '';
+  for (var modifier in Blockly.KeyboardShortcutRegistry.modifierKeys) {
+    if (e.getModifierState(modifier)) {
+      if (serializedKey != '') {
+        serializedKey += '+';
+      }
+      serializedKey += modifier;
+    }
+  }
+  if (serializedKey != '' && e.keyCode) {
+    serializedKey = serializedKey + '+' + e.keyCode;
+  } else if (e.keyCode) {
+    serializedKey = e.keyCode.toString();
+  }
+  return serializedKey;
+};
+
+/**
+ * Checks whether any of the given modifiers are not valid.
+ * @param {!Array<string>} modifiers List of modifiers to be used with the key.
+ * @throws {Error} if the modifier is not in the valid modifiers list.
+ * @private
+ */
+Blockly.KeyboardShortcutRegistry.prototype.checkModifiers_ = function(
+    modifiers) {
+  var validModifiers = Blockly.utils.object.values(
+      Blockly.KeyboardShortcutRegistry.modifierKeys);
+  for (var i = 0, modifier; (modifier = modifiers[i]); i++) {
+    if (validModifiers.indexOf(modifier) < 0) {
+      throw Error(modifier + ' is not a valid modifier key.');
```

Comment:
Always throw with `new` according to the styleguide: https://google.github.io/styleguide/jsguide.html#features-exceptions

but there's a number of instances of this in Blockly, so happy to keep it optional.

Reply:
done.